main/pipeline/pipeline.cc

```diff
@@ -428,55 +428,107 @@ struct IndexThreadResultPack {
+            this->subst.emplace(*this->threadGs, to);
+        } else {
+            core::GlobalSubstitution::mergeFileTables(*this->threadGs, to);
+        }
+    }
+
+    IndexSubstitutionJob(IndexSubstitutionJob &&other) = default;
+    IndexSubstitutionJob &operator=(IndexSubstitutionJob &&other) = default;
+};
+
+vector<ast::ParsedFile> mergeIndexResults(core::GlobalState &cgs, const options::Options &opts,
+                                          shared_ptr<BlockingBoundedQueue<IndexThreadResultPack>> input,
+                                          WorkerPool &workers, const unique_ptr<const OwnedKeyValueStore> &kvstore) {
     ProgressIndicator progress(opts.showProgress, "Indexing", input->bound);
-    Timer timeit(cgs->tracer(), "mergeIndexResults");
-    IndexThreadResultPack threadResult;
-    IndexResult ret;
-    for (auto result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs->tracer()); !result.done();
-         result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs->tracer())) {
-        if (result.gotItem()) {
-            counterConsume(move(threadResult.counters));
-            if (ret.gs == nullptr) {
-                ret.gs = move(threadResult.res.gs);
-                ENFORCE(ret.trees.empty());
-                ret.trees = move(threadResult.res.trees);
-            } else {
-                core::GlobalSubstitution substitution(*threadResult.res.gs, *ret.gs, cgs.get());
-                {
-                    Timer timeit(cgs->tracer(), "substituteTrees");
-                    for (auto &tree : threadResult.res.trees) {
-                        auto file = tree.file;
-                        if (!file.data(*ret.gs).cached) {
-                            core::MutableContext ctx(*ret.gs, core::Symbols::root(), file);
-                            tree.tree = ast::Substitute::run(ctx, substitution, move(tree.tree));
+    Timer timeit(cgs.tracer(), "mergeIndexResults");
+
+    auto batchq = make_shared<ConcurrentBoundedQueue<IndexSubstitutionJob>>(input->bound);
+    vector<ast::ParsedFile> ret;
+
+    {
+        Timer timeit(cgs.tracer(), "mergeGlobalStates");
+        IndexThreadResultPack threadResult;
+        for (auto result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs.tracer());
+             !result.done(); result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs.tracer())) {
+            if (result.gotItem()) {
+                counterConsume(move(threadResult.counters));
+                auto numTrees = threadResult.res.trees.size();
+                batchq->push(IndexSubstitutionJob{cgs, std::move(threadResult.res)}, numTrees);
+            }
+        }
+    }
+
+    {
+        Timer timeit(cgs.tracer(), "substituteTrees");
+        auto resultq = make_shared<BlockingBoundedQueue<vector<ast::ParsedFile>>>(batchq->bound);
+
+        workers.multiplexJob("substituteTrees", [&cgs, batchq, resultq]() {
```

Comment:
Can we explicitly pass a `const` reference to GlobalState to this lambda to prevent accidental mutation?

Reply:
Probably, I'll investigate.

Reply:
Since `Substitute` requires a `core::MutableContext`, this can't be made `const`. Maybe if we could do templates to `Substitute` to make that work.